README.md
# FakeQL

Automatic GraphQL mocks

## Installation

### npm

```sh
npm install --save-dev fakeql
```

### yarn

```sh
yarn add --dev fakeql
```

## Usage

### Basic

```javascript
import { fakeQL } from "fakeql";
import { parse, buildSchema } from "graphql";

const schema = buildSchema(`
  type Query {
    id: ID!
    string: String!
    int: Int!
    float: Float!
    boolean: Boolean!
  }
`);
const document = parse(`
  query {
    __typename
    id
    string
    int
    float
    boolean
  }
`);

fakeQL({ document, schema });
/* { 
  __typename: "Query",
  id: `mock-value-for-field-"id"`,
  string: `mock-value-for-field-"string"`,
  int: 42,
  float: 4.2,
  boolean: false,
}
*/
```

### graphql-config

If no `schema` is provided FakeQL will look for one using [graphql-config]. Assuming it is configured the above example would become:

```javascript
import { fakeQL } from "fakeql";
import { parse } from "graphql";

const document = parse(`
  query {
    __typename
    id
    string
    int
    float
    boolean
  }
`);

fakeQL({ document });
/* { 
  __typename: "Query",
  id: `mock-value-for-field-"id"`,
  string: `mock-value-for-field-"string"`,
  int: 42,
  float: 4.2,
  boolean: false,
}
*/
```

[graphql-config]: https://github.com/kamilkisiela/graphql-config/tree/legacy#usage

### Resolvers

Real life queries will be more complex and in real testing you'll want to configure specific parts of the mock. Each mock can be fine tuned by passing a `resolvers` map to `fakeQL`. For example to change the default value of the `String` and `Int` scalars:

```javascript
const document = parse(`
  query me {
    me { # User!
      name # String!
      age # Int!
    }
  }
`);

fakeQL({
  document,
  resolvers: {
    String() {
      return "custom-string";
    },
    Int() {
      return 84;
    },
  },
})
/* {
  me: {
    name: "custom-string",
    age: 84,
  },
}
*/
```

This also works for types and custom enums:

```javascript
const schema = buildSchema(`
  enum Role {
    ADMIN # By default fakeQL picks this first enum value
    MEMBER
  }

  type User {
    name: String!
    role: Role!
    age: Int!
  }

  type Query {
    me: User
  }
`);
const document = parse(`
  query {
    me {
      name
      age
      role
    }
  }
`);

fakeQL({
  document,
  schema,
  resolvers: {
    User() {
      return { name: "Hello" }
    }
    Role() {
      return "MEMBER";
    },
  },
});
/* {
  me: {
    name: "Hello",
    age: 42,
    role: "MEMBER",
  }
}
*/
```


CHANGELOG.md
# Unreleased

- Add support for nested fragments

# 0.1.0 (2020-03-31)

- Add MIT license
- Add support for __typename meta field
- Add default values for all standard scalars
- Add support for custom scalar, type and enum resolvers
- Add support for fragments


# 0.0.2 (2020-03-25)

Initial release



src/fakeql.test.ts
import { parse, buildSchema, introspectionFromSchema } from "graphql";
import { fakeQL, FakeQLError } from "./index";

import { writeFileSync, unlinkSync } from "fs";

describe("fakeQL", () => {
  it("mocks scalars with default values", () => {
    const schema = buildSchema(`
      type Query {
        id: ID!
        nullId: ID
        string: String!
        nullString: String
        int: Int!
        nullInt: Int
        float: Float!
        nullFloat: Float
        boolean: Boolean!
        nullBoolean: Boolean
      }
    `);
    const document = parse(`
      query {
        __typename
        id
        nullId
        string
        nullString
        int
        nullInt
        float
        nullFloat
        boolean
        nullBoolean
      }
    `);

    const mock = fakeQL({ document, schema });

    expect(mock).toEqual({
      __typename: "Query",
      id: `mock-value-for-field-"id"`,
      nullId: `mock-value-for-field-"nullId"`,
      string: `mock-value-for-field-"string"`,
      nullString: `mock-value-for-field-"nullString"`,
      int: 42,
      nullInt: 42,
      float: 4.2,
      nullFloat: 4.2,
      boolean: false,
      nullBoolean: false,
    });
  });

  it("mocks custom types", () => {
    const schema = buildSchema(`
      type User {
        name: String!
        age: Int!
      }

      type Query {
        me: User
      }
    `);
    const document = parse(`
      query {
        __typename
        me {
          __typename
          name
          age
        }
      }
    `);

    const mock = fakeQL({ document, schema });

    expect(mock).toEqual({
      __typename: "Query",
      me: {
        __typename: "User",
        name: `mock-value-for-field-"name"`,
        age: 42,
      },
    });
  });

  it("mocks lists with scalars", () => {
    const schema = buildSchema(`
      type User {
        hobbies: [String]!
      }

      type Query {
        me: User
      }
    `);
    const document = parse(`
      query {
        __typename
        me {
          __typename
          hobbies
        }
      }
    `);

    const mock = fakeQL({ document, schema });

    expect(mock).toEqual({
      __typename: "Query",
      me: {
        __typename: "User",
        hobbies: [`mock-value-for-field-"hobbies"`],
      },
    });
  });

  it("mocks lists with enums", () => {
    const schema = buildSchema(`
      type User {
        roles: [Role]!
      }

      enum Role {
        ADMIN
        MEMBER
      }

      type Query {
        me: User
      }
    `);
    const document = parse(`
      query {
        me {
          roles
        }
      }
    `);

    const mock = fakeQL({ document, schema });

    expect(mock).toEqual({
      me: {
        roles: ["ADMIN"],
      },
    });
  });

  it("mocks lists with custom types", () => {
    const schema = buildSchema(`
      type User {
        hobbies: [Hobby]!
      }

      type Hobby {
        name: String!
      }

      type Query {
        me: User
      }
    `);
    const document = parse(`
      query {
        __typename
        me {
          __typename
          hobbies {
            __typename
            name
          }
        }
      }
    `);

    const mock = fakeQL({ document, schema });

    expect(mock).toEqual({
      __typename: "Query",
      me: {
        __typename: "User",
        hobbies: [{ __typename: "Hobby", name: `mock-value-for-field-"name"` }],
      },
    });
  });

  it("mocks weirdly nested lists", () => {
    const schema = buildSchema(`
      type User {
        name: String!
        hobbies: [[[String!]!]!]
        friends: [[[User!]!]!]!
      }


      type Query {
        me: User
      }
    `);
    const document = parse(`
      query {
        __typename
        me {
          __typename
          hobbies
          friends {
            __typename
            name
          }
        }
      }
    `);

    const mock = fakeQL({ document, schema });

    expect(mock).toEqual({
      __typename: "Query",
      me: {
        __typename: "User",
        hobbies: [[[`mock-value-for-field-"hobbies"`]]],
        friends: [
          [[{ __typename: "User", name: `mock-value-for-field-"name"` }]],
        ],
      },
    });
  });

  it("mocks custom enums", () => {
    const schema = buildSchema(`
      type User {
        name: String!
        role: Role!
      }

      enum Role {
        ADMIN
        MEMBER
      }

      type Query {
        me: User
      }
    `);
    const document = parse(`
      query {
        me {
          name
          role
        }
      }
    `);

    const mock = fakeQL({ document, schema });

    expect(mock).toEqual({
      me: { name: `mock-value-for-field-"name"`, role: "ADMIN" },
    });
  });

  it("mocks nested types", () => {
    const schema = buildSchema(`
      type User {
        name: String!
        friends: [User]!
        age: Int!
        role: String!
      }

      type Query {
        me: User
      }
    `);
    const document = parse(`
      query {
        __typename
        me {
          __typename
          friends {
            __typename
            name
            age
            friends {
              friends {
                name
              }
            }
          }
          # Order is important. Putting role below friends ensures that we enter
          # the friends list generete the mock friend and then leave the list to
          # generate the mock role.
          role
        }
      }
    `);

    const mock = fakeQL({ document, schema });

    expect(mock).toEqual({
      __typename: "Query",
      me: {
        __typename: "User",
        role: `mock-value-for-field-"role"`,
        friends: [
          {
            __typename: "User",
            name: `mock-value-for-field-"name"`,
            age: 42,
            friends: [
              {
                friends: [
                  {
                    name: `mock-value-for-field-"name"`,
                  },
                ],
              },
            ],
          },
        ],
      },
    });
  });

  it("mocks sibling types", () => {
    const schema = buildSchema(`
      type User {
        name: String!
        friends: [User]!
        invoices: [Invoice]!
        age: Int!
        role: String!
      }

      type Invoice {
        date: String!
      }

      type Query {
        me: User
      }
    `);
    const document = parse(`
      query {
        __typename
        me {
          __typename
          friends {
            __typename
            name
            age
          }
          invoices {
            __typename
            date
          }
        }
      }
    `);

    const mock = fakeQL({ document, schema });

    expect(mock).toEqual({
      __typename: "Query",
      me: {
        __typename: "User",
        friends: [
          { __typename: "User", name: `mock-value-for-field-"name"`, age: 42 },
        ],
        invoices: [
          { __typename: "Invoice", date: `mock-value-for-field-"date"` },
        ],
      },
    });
  });

  it("mocks nested and sibling types", () => {
    const schema = buildSchema(`
      type User {
        name: String!
        friends: [User]!
        invoices: [Invoice]!
        age: Int!
        role: String!
      }

      type Invoice {
        date: String!
        user: User!
      }

      type Query {
        me: User
      }
    `);
    const document = parse(`
      query {
        __typename
        me {
          __typename
          friends {
            __typename
            name
            invoices {
              user {
                invoices {
                  date
                  user {
                    name
                  }
                }
                name
              }
            }
            age
          }
          invoices {
            __typename
            date
            user {
              invoices {
                user {
                  name
                }
                date
              }
            }
          }
        }
      }
    `);

    const mock = fakeQL({ document, schema });

    expect(mock).toEqual({
      __typename: "Query",
      me: {
        __typename: "User",
        friends: [
          {
            __typename: "User",
            name: `mock-value-for-field-"name"`,
            age: 42,
            invoices: [
              {
                user: {
                  name: `mock-value-for-field-"name"`,
                  invoices: [
                    {
                      date: `mock-value-for-field-"date"`,
                      user: {
                        name: `mock-value-for-field-"name"`,
                      },
                    },
                  ],
                },
              },
            ],
          },
        ],
        invoices: [
          {
            __typename: "Invoice",
            date: `mock-value-for-field-"date"`,
            user: {
              invoices: [
                {
                  date: `mock-value-for-field-"date"`,
                  user: { name: `mock-value-for-field-"name"` },
                },
              ],
            },
          },
        ],
      },
    });
  });

  it("mocks fragments (fragment comes first)", () => {
    const schema = buildSchema(`
      type User {
        name: String!
      }

      type Query {
        me: User!
      }
    `);
    const document = parse(`
      fragment user on User {
        name
      }

      query {
        me {
          ...user
        }
      }
    `);

    const mock = fakeQL({ document, schema });

    expect(mock).toEqual({
      me: { name: `mock-value-for-field-"name"` },
    });
  });

  it("mocks fragments (fragment comes last)", () => {
    const schema = buildSchema(`
      type User {
        name: String!
      }

      type Query {
        me: User!
      }
    `);
    const document = parse(`
      query {
        me {
          ...user
        }
      }

      fragment user on User {
        name
      }
    `);

    const mock = fakeQL({ document, schema });

    expect(mock).toEqual({
      me: { name: `mock-value-for-field-"name"` },
    });
  });

  it("mocks sibling fragments", () => {
    const schema = buildSchema(`
      type User {
        name: String!
        age: Int!
      }

      type Query {
        me: User!
      }
    `);
    const document = parse(`
      query {
        me {
          ...user
        }
      }

      fragment user on User {
        ...name
        ...age
      }

      fragment name on User {
        name
      }

      fragment age on User {
        age
      }
    `);

    const mock = fakeQL({ document, schema });

    expect(mock).toEqual({
      me: {
        name: `mock-value-for-field-"name"`,
        age: 42,
      },
    });
  });

  it("mocks nested fragments", () => {
    const schema = buildSchema(`
      type User {
        age: Int!
        hobbies: [Hobby!]!
      }

      type Hobby {
        name: String!
      }

      type Query {
        me: User!
      }
    `);
    const document = parse(`
      query {
        me {
          ...user
        }
      }

      fragment user on User {
        ...hobbies
        age
      }

      fragment hobbies on User {
        hobbies {
          name
        }
      }
    `);

    const mock = fakeQL({ document, schema });

    expect(mock).toEqual({
      me: {
        hobbies: [{ name: `mock-value-for-field-"name"` }],
        age: 42,
      },
    });
  });

  it("mocks nested list fragments", () => {
    const schema = buildSchema(`
      type User {
        hobbies: [Hobby!]!
      }

      type Hobby {
        name: String!
      }

      type Query {
        me: User!
      }
    `);
    const document = parse(`
      query {
        me {
          ...user
        }
      }

      fragment user on User {
        hobbies {
          ...hobby
        }
      }

      fragment hobby on Hobby {
        name
      }
    `);

    const mock = fakeQL({ document, schema });

    expect(mock).toEqual({
      me: {
        hobbies: [{ name: `mock-value-for-field-"name"` }],
      },
    });
  });

  it("merges fragments with existing mocks", () => {
    const schema = buildSchema(`
      type User {
        id: ID!
        name: String!
        age: Int!
      }

      type Query {
        me: User!
      }
    `);
    const document = parse(`
      query {
        me {
          name
          ...user
          age
        }
      }

      fragment user on User {
        id
      }
    `);

    const mock = fakeQL({ document, schema });

    expect(mock).toEqual({
      me: {
        id: `mock-value-for-field-"id"`,
        name: `mock-value-for-field-"name"`,
        age: 42,
      },
    });
  });

  it("throws when spreading an unknown fragment", () => {
    const schema = buildSchema(`
      type User {
        name: String!
      }

      type Query {
        me: User!
      }
    `);
    const document = parse(`
      query {
        me {
          ...unknown
        }
      }
    `);

    expect(() => {
      fakeQL({ document, schema });
    }).toThrow();
  });

  it("creates the schema from an IntrospectionQuery", () => {
    const schema = buildSchema(`
      type User {
        name: String
      }

      type Query {
        me: User!
      }
    `);
    const document = parse(`
      query me {
        me {
          name
        }
      }
    `);
    const introspectionQuery = introspectionFromSchema(schema);

    expect(
      fakeQL({
        document,
        schema: introspectionQuery,
      })
    ).toEqual({
      me: {
        name: 'mock-value-for-field-"name"',
      },
    });
  });

  it("supports custom scalar resolvers", () => {
    const schema = buildSchema(`
      type User {
        name: String!
        age: Int!
      }

      type Query {
        me: User!
      }
    `);
    const document = parse(`
      query me {
        me {
          name
          age
        }
      }
    `);
    const introspectionQuery = introspectionFromSchema(schema);

    expect(
      fakeQL({
        document,
        schema: introspectionQuery,
        resolvers: {
          String(): string {
            return "custom-string";
          },
          Int(): number {
            return 84;
          },
        },
      })
    ).toEqual({
      me: {
        name: "custom-string",
        age: 84,
      },
    });
  });

  it("supports custom type resolvers", () => {
    const schema = buildSchema(`
      type User {
        name: String!
        age: Int!
      }
      type Query {
        me: User!
      }
    `);
    const document = parse(`
      query me {
        me {
          name
          age
        }
      }
    `);
    const introspectionQuery = introspectionFromSchema(schema);

    expect(
      fakeQL({
        document,
        schema: introspectionQuery,
        resolvers: {
          User(): unknown {
            return { name: "Hello" };
          },
        },
      })
    ).toEqual({
      me: {
        name: "Hello",
        age: 42,
      },
    });
  });

  it("supports custom enum type resolvers", () => {
    const schema = buildSchema(`
      type User {
        name: String!
        role: Role!
      }

      enum Role {
        ADMIN
        MEMBER
      }

      type Query {
        me: User
      }
    `);
    const document = parse(`
      query {
        me {
          name
          role
        }
      }
    `);

    const mock = fakeQL({
      document,
      schema,
      resolvers: {
        Role(): string {
          return "MEMBER";
        },
      },
    });

    expect(mock).toEqual({
      me: { name: `mock-value-for-field-"name"`, role: "MEMBER" },
    });
  });

  it("fails when schema is invalid", () => {
    const schema = buildSchema(`
      # Schema is missing type Query
      type Team {
        name: String!
        userCanAdminister: Boolean!
      }
    `);

    expect(() => {
      fakeQL({
        document: parse(`query { me { name } }`),
        schema,
      });
    }).toThrow();
  });

  it("fails when document is invalid", () => {
    const schema = buildSchema(`type Query { x: String! }`);
    const document = parse(`
      # Document is querying a type that doesn't exit
      type Team {
        name: String!
        userCanAdminister: Boolean!
      }
    `);

    expect(() =>
      fakeQL({
        document,
        schema,
      })
    ).toThrow();
  });

  it("supports customizing the validation rules", () => {
    const schema = buildSchema(`type Query { x: String! }`);
    const document = parse(`
      # This is a syntactically valid document but querying the wrong fields
      # With all validation rules disabled querying the wrong fields won't throw
      # an error
      type Team {
        name: String!
        userCanAdminister: Boolean!
      }
    `);

    expect(() =>
      fakeQL({
        document,
        schema,
        validationRules: [],
      })
    ).not.toThrow();
  });

  it("can take schema from graphql-config", () => {
    writeFileSync(
      "./.graphqlconfig",
      JSON.stringify({ schemaPath: "./schema.json" })
    );
    writeFileSync(
      "./schema.json",
      JSON.stringify(
        introspectionFromSchema(
          buildSchema(`
          type Query { x: String ! }
        `)
        ),
        null,
        2
      )
    );
    const document = parse(`query { x }`);

    const mock = fakeQL({ document });

    expect(mock).toEqual({ x: `mock-value-for-field-"x"` });

    unlinkSync("./schema.json");
    unlinkSync("./.graphqlconfig");
  });

  it("fails when no schema is provided and graphql-config cannot be found", () => {
    const document = parse(`query { x }`);

    expect(() => {
      fakeQL({ document });
    }).toThrow(FakeQLError);
  });
});


src/assign/assign.test.ts
import { assign } from "./index";

describe("assign", () => {
  it("returns unmodified object when path is empty", () => {
    expect(assign({ name: "Wei Shi Lindon" }, [], "")).toEqual({
      name: "Wei Shi Lindon",
    });
  });

  it("assigns to the value at path", () => {
    expect(assign({ name: "" }, ["name"], "Wei Shi Lindon")).toEqual({
      name: "Wei Shi Lindon",
    });
  });

  it("works with nested paths", () => {
    expect(assign({ me: {} }, ["me", "name"], "Wei Shi Lindon")).toEqual({
      me: {
        name: "Wei Shi Lindon",
      },
    });
  });

  it("creates the nested path if undefined", () => {
    expect(assign({}, ["me", "name"], "Wei Shi Lindon")).toEqual({
      me: {
        name: "Wei Shi Lindon",
      },
    });
  });

  it("creates the nested path if null", () => {
    expect(assign({ me: null }, ["me", "name"], "Wei Shi Lindon")).toEqual({
      me: {
        name: "Wei Shi Lindon",
      },
    });
  });

  it("supports array indicies", () => {
    expect(
      assign({ me: { teams: [] } }, ["me", "teams", 0, "name"], "Arelius")
    ).toEqual({
      me: {
        teams: [{ name: "Arelius" }],
      },
    });
  });
});


src/fakeql.ts
import {
  visit,
  visitWithTypeInfo,
  isNonNullType,
  DocumentNode,
  validate,
  validateSchema,
  TypeInfo,
  GraphQLSchema,
  isObjectType,
  isScalarType,
  GraphQLScalarType,
  isListType,
  GraphQLCompositeType,
  buildClientSchema,
  isSchema,
  IntrospectionQuery,
  ValidationRule,
  isEnumType,
  ASTNode,
  VisitFn,
  isWrappingType,
  FragmentDefinitionNode,
  GraphQLEnumType,
} from "graphql";
import { FakeQLError } from "./error";
import { getGraphQLProjectConfig } from "graphql-config";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Mock = Record<string, any>;

type MockResolverMap = { [key: string]: MockResolver };
interface MockResolver {
  (): unknown;
}

const valueForScalarType = (
  type: GraphQLScalarType,
  name: string,
  resolvers?: MockResolverMap
): unknown => {
  if (resolvers && resolvers[type.name]) {
    return resolvers[type.name]();
  }

  switch (type.name) {
    default:
    case "String":
      return `mock-value-for-field-"${name}"`;

    case "Int":
      return 42;

    case "Float":
      return 4.2;

    case "Boolean":
      return false;
  }
};

const valueForEnumType = (
  type: GraphQLEnumType,
  resolvers: MockResolverMap
): string | undefined => {
  const resolver = resolvers[type.name];

  if (resolver) {
    return resolver() as string;
  }

  const values = type.getValues();
  // A GraphQL enum with no members should be impossible, but it's
  // not enforced in the types so we double check it here.
  if (values.length > 0) {
    return values[0].name;
  }
};

interface FakeReducerArguments {
  ast: ASTNode;
  typeInfo: TypeInfo;
  resolvers: MockResolverMap;
  fragments: {
    [key: string]: { definition: FragmentDefinitionNode; mock?: Mock };
  };
}
const fakeFor = ({
  ast,
  typeInfo,
  resolvers,
  fragments,
}: FakeReducerArguments): Mock => {
  const notHandled: VisitFn<ASTNode, ASTNode> = () => null;

  return visit(
    ast,
    visitWithTypeInfo(typeInfo, {
      leave: {
        Document(node): Mock {
          return node.definitions[0];
        },

        OperationDefinition(node): Mock {
          return node.selectionSet;
        },

        SelectionSet(node): Mock {
          let type = typeInfo.getType();

          if (isNonNullType(type)) {
            type = type.ofType;
          }

          const object = node.selections.reduce(
            (mock, selection) => ({ ...mock, ...selection }),
            {}
          );

          return object;
        },

        FragmentSpread(node): Mock {
          const fragment = fragments[node.name.value];

          if (!fragment.mock) {
            fragment.mock = fakeFor({
              ast: fragment.definition.selectionSet,
              typeInfo,
              resolvers,
              fragments,
            });
          }

          return fragment.mock;
        },

        Field(node): Mock | null {
          let type = typeInfo.getType();

          if (isNonNullType(type)) {
            type = type.ofType;
          }

          if (node.name.value === "__typename" && typeInfo.getParentType()) {
            const parentType = typeInfo.getParentType() as GraphQLCompositeType;
            return { __typename: parentType.name };
          } else if (isScalarType(type)) {
            return {
              [node.name.value]: valueForScalarType(
                type,
                node.name.value,
                resolvers
              ),
            };
          } else if (isObjectType(type)) {
            const resolver = resolvers[type.name];
            if (resolver) {
              const resolvedValue = resolver() as Mock;
              return {
                [node.name.value]: { ...node.selectionSet, ...resolvedValue },
              };
            }

            return { [node.name.value]: node.selectionSet };
          } else if (isListType(type)) {
            let itemType = type.ofType;

            let depth = 1;

            while (isWrappingType(itemType)) {
              if (isListType(itemType)) {
                depth += 1;
              }
              itemType = itemType.ofType;
            }

            let value: unknown = node.selectionSet;
            if (isScalarType(itemType)) {
              value = valueForScalarType(itemType, node.name.value, resolvers);
            } else if (isEnumType(itemType)) {
              value = valueForEnumType(itemType, resolvers);
            }

            while (depth > 0) {
              depth -= 1;

              value = [value];
            }

            return {
              [node.name.value]: value,
            };

            return { [node.name.value]: node.selectionSet };
          } else if (isEnumType(type)) {
            return { [node.name.value]: valueForEnumType(type, resolvers) };
          }

          return null;
        },
        Name(node): ASTNode {
          return node;
        },

        VariableDefinition: notHandled,
        Variable: notHandled,
        Argument: notHandled,
        InlineFragment: notHandled,
        FragmentDefinition: notHandled,
        IntValue: notHandled,
        FloatValue: notHandled,
        StringValue: notHandled,
        BooleanValue: notHandled,
        NullValue: notHandled,
        EnumValue: notHandled,
        ListValue: notHandled,
        ObjectValue: notHandled,
        ObjectField: notHandled,
        Directive: notHandled,
        NamedType: notHandled,
        ListType: notHandled,
        NonNullType: notHandled,
        SchemaDefinition: notHandled,
        OperationTypeDefinition: notHandled,
        ScalarTypeDefinition: notHandled,
        ObjectTypeDefinition: notHandled,
        FieldDefinition: notHandled,
        InputValueDefinition: notHandled,
        InterfaceTypeDefinition: notHandled,
        UnionTypeDefinition: notHandled,
        EnumTypeDefinition: notHandled,
        EnumValueDefinition: notHandled,
        InputObjectTypeDefinition: notHandled,
        DirectiveDefinition: notHandled,
        SchemaExtension: notHandled,
        ScalarTypeExtension: notHandled,
        ObjectTypeExtension: notHandled,
        InterfaceTypeExtension: notHandled,
        UnionTypeExtension: notHandled,
        EnumTypeExtension: notHandled,
        InputObjectTypeExtension: notHandled,
      },
    })
  );
};

export interface FakeQLArguments {
  document: DocumentNode;
  schema?: GraphQLSchema | IntrospectionQuery;
  resolvers?: MockResolverMap;
  validationRules?: ValidationRule[];
}
export const fakeQL = ({
  document,
  schema,
  resolvers = {},
  validationRules,
}: FakeQLArguments): Mock => {
  if (schema && !isSchema(schema)) {
    schema = buildClientSchema(schema);
  }

  if (!schema) {
    try {
      schema = getGraphQLProjectConfig().getSchema();
    } catch (error) {
      throw new FakeQLError(
        "There was a problem with your graphql-config and no schema was provided as an argument",
        [error]
      );
    }
  }

  const schemaValidationErrors = validateSchema(schema);
  if (schemaValidationErrors.length > 0) {
    throw new FakeQLError("Invalid Schema", schemaValidationErrors);
  }

  const validationErrors = validate(schema, document, validationRules);
  if (validationErrors.length > 0) {
    throw new FakeQLError("Invalid Document", validationErrors);
  }

  const typeInfo = new TypeInfo(schema);

  const fragments: {
    [key: string]: { definition: FragmentDefinitionNode; mock?: Mock };
  } = {};

  for (const definition of document.definitions) {
    if (definition.kind !== "FragmentDefinition") {
      continue;
    }

    fragments[definition.name.value] = { definition };
  }

  return fakeFor({ ast: document, typeInfo, resolvers, fragments });
};
